File: spring-graphql-webflux/src/main/java/com/example/demo/service/AuthorService.java
package com.example.demo.service;

import com.example.demo.gql.types.Author;
import com.example.demo.gql.types.Post;
import com.example.demo.model.AuthorEntity;
import com.example.demo.model.PostEntity;
import com.example.demo.repository.AuthorRepository;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import reactor.core.publisher.Flux;
import reactor.core.publisher.Mono;

import java.util.*;
import java.util.function.Function;

@Service
@RequiredArgsConstructor
public class AuthorService {
    public static final Function<AuthorEntity, Author> MAPPER = entity -> Author.builder()
            .id(entity.id().toString())
            .name(entity.name())
            .email(entity.email())
            .build();

    private final AuthorRepository authors;

    public Mono<Author> getAuthorById(String id) {
        return this.authors.findById(UUID.fromString(id))
                .map(MAPPER)
                .switchIfEmpty(Mono.error(new AuthorNotFoundException(id)));
    }

    public Flux<Author> getAuthorByIdIn(List<String> keys) {
        List<UUID> uuids = keys.stream().map(UUID::fromString).toList();
        return this.authors.findByIdIn(uuids).map(MAPPER);
    }
}

File: graphql-kotlin/src/main/kotlin/com/example/demo/service/AuthorService.kt
package com.example.demo.service

import com.example.demo.asGqlType
import com.example.demo.gql.types.Author
import com.example.demo.repository.AuthorRepository
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.map
import org.springframework.stereotype.Service
import java.util.*

@Service
class AuthorService(val authors: AuthorRepository) {

    suspend fun getAuthorById(id: UUID): Author {
        val author = this.authors.findById(id) ?: throw AuthorNotFoundException(id)
        return author.asGqlType()
    }

    // alternative to use kotlin `Flow`
    fun getAuthorByIdIn(ids: List<UUID>): Flow<Author> {
        return authors.findAllById(ids.toList()).map { it.asGqlType() }
    }
}